torch/distributed/_cuda_p2p/__init__.py

```diff
@@ -121,3 +121,360 @@ def get_p2p_buffer_size(group: c10d.ProcessGroup) -> int:
+    ag_out: torch.Tensor,
+    group: c10d.ProcessGroup,
+) -> None:
+    """
+    Perform the following logic with micro-pipelined computation and
+    communication:
+
+        tensor = all_gather_tensor(shard, gather_dim=1, group=group)
+        chunks = tensor.chunk(group.size())
+        for src_rank, chunk in enumerate(chunks):
+            shard_consumer(chunk, src_rank)
+
+    NOTE:
+    - The shard passed to shard consumer will always be contiguous.
+    """
+    p2p_buf_sz_req = shard.numel() * shard.element_size()
+    if get_p2p_buffer_size(group) < p2p_buf_sz_req:
+        # We preferred the caller to handle fallback so that the computation
+        # doesn't need to be decomposed.
+        raise RuntimeError(
+            f"_pipelined_all_gather_and_consume on input with shape={shard.shape} "
+            f"and dtype={shard.dtype} requires {p2p_buf_sz_req} bytes of p2p buffers "
+            f"(got {get_p2p_buffer_size(group)} bytes)."
+        )
+
+    backend = get_cuda_p2p_backend(group)
+    group_size = group.size()
+    rank = group.rank()
+
+    backend.stream().wait_stream(torch.cuda.current_stream())
+    local_p2p_buf = backend.get_p2p_buffer(rank, shard.shape, shard.dtype)
+
+    chunks = ag_out.chunk(group.size())
+
+    # While consuming local shard, copy it to the local p2p buffer
+    # in another stream.
+    shard_consumer(shard, rank)
+    chunks[rank].copy_(shard)
+
+    with torch.cuda.stream(backend.stream()):
+        local_p2p_buf.copy_(shard)
+        work = backend.intra_node_barrier()
+    work.wait()
+
+    # At this point, all ranks have copied their local shard to
+    # their local p2p buffer. Each rank can now copy and consume
+    # remote shards.
+    for i in range(1, group_size):
+        if i % 2 == 0:
+            stream = torch.cuda.current_stream()
+        else:
+            stream = backend.stream()
+        remote_rank = (i + rank) % group_size
+        remote_p2p_buf = backend.get_p2p_buffer(remote_rank, shard.shape, shard.dtype)
+        with torch.cuda.stream(stream):
+            chunks[remote_rank].copy_(remote_p2p_buf)
+            shard_consumer(chunks[remote_rank], remote_rank)
+
+    torch.cuda.current_stream().wait_stream(backend.stream())
+
+    with torch.cuda.stream(backend.stream()):
+        work = backend.intra_node_barrier()
+    work.wait()
+
+
+def _pipelined_produce_and_all2all(
+    chunk_producer: Callable[[int, torch.Tensor], None],
+    output: torch.Tensor,
+    group: c10d.ProcessGroup,
+) -> None:
+    """
+    Perform the following logic with micro-pipelined computation and
+    communication:
+
+        chunks = [
+            chunk_producer(dst_rank, chunks[dst_rank])
+            for dst_rank in range(group.size()):
+        ]
+        dist.all_to_all_single(output=output, input=torch.cat(chunks))
+    """
+    group_size = group.size()
+    rank = group.rank()
+
+    out_chunks = output.chunk(group_size)
```

Comment:
Does this approach in higher memory usage? Since we need to materialize both the input and output at the same time?

Reply:
It is higher compared to the online accumulation approach where each rank only needs to maintain the `reduce_scatter` result. It is not higher compare to the baseline which materializes the matmul output anyway.

We do require a p2p buffer 2x the size of the `reduce_scatter` result to allow for more parallelism.